File: apps/lex-core-service/src/main/java/com/infosys/lex/interest/service/InterestService.java
/*               "Copyright 2020 Infosys Ltd.
               Use of this source code is governed by GPL v3 license that can be found in the LICENSE file or at https://opensource.org/licenses/GPL-3.0
               This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License version 3" */
substitute url based on requirement

import java.util.List;
import java.util.Map;

import javax.validation.constraints.NotNull;

public interface InterestService {

	/**
	 * get interests of users
	 * 
	 * @param rootOrg
	 * @param userId
	 * @return
	 * @throws Exception
	 */
	public Map<String, Object> getInterest(String rootOrg, String userId) throws Exception;

	/**
	 * delete interests of user
	 * 
	 * @param rootOrg
	 * @param userId
	 * @param interest
	 * @return
	 * @throws Exception
	 */
	public String delete(String rootOrg, String userId, String interest) throws Exception;

	/**
	 * add or create interest
	 * 
	 * @param rootOrg
	 * @param userId
	 * @param interest
	 * @return
	 * @throws Exception
	 */
	public String upsert(String rootOrg, @NotNull String userId, @NotNull String interest) throws Exception;

	/**
	 * autocompletes users interests
	 * 
	 * @param rootOrg
	 * @param org
	 * @param language
	 * @param query
	 * @param type
	 * @return
	 * @throws Exception
	 */
	public List<String> autoComplete(String rootOrg, String org, @NotNull String language, String query, String type)
			throws Exception;

	/**
	 * get suggested interests
	 * 
	 * @param rootOrg
	 * @param userid
	 * @param org
	 * @param language
	 * @return
	 * @throws Exception
	 */

	public List<String> suggestedComplete(String rootOrg, String userid, String org, @NotNull String language)
			throws Exception;
}

File: apps/fusion/project/ws/app/src/lib/routes/profile/routes/interest/services/interest.service.ts
/*               "Copyright 2020 Infosys Ltd.
               Use of this source code is governed by GPL v3 license that can be found in the LICENSE file or at https://opensource.org/licenses/GPL-3.0
               This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License version 3" */
import { Injectable } from '@angular/core'
import { Observable } from 'rxjs'
// import { WsApiCatalogService } from '@ws-shared/apis/src/lib/ws-api-catalog.service'
// import { WsSharedUtilityService } from '@ws-shared/services'
import { HttpClient } from '@angular/common/http'
import { map } from 'rxjs/operators'
import { ITopic, ITopicRecommended } from '../models/interest.model'
import { NSSearch, TreeCatalogService } from '@ws-widget/collection'
const PROTECTED_SLAG_V8 = `/apis/protected/v8`
const API_END_POINTS = {
  userTopics: `${PROTECTED_SLAG_V8}/user/topics`,
  recommendedTopics: `${PROTECTED_SLAG_V8}/user/topics/suggested`,
  autocompleteTopics: `${PROTECTED_SLAG_V8}/user/topics/autocomplete`,
}
@Injectable({
  providedIn: 'root',
})
export class InterestService {
  constructor(private http: HttpClient, private treeCatalogSvc: TreeCatalogService) {}
  fetchUserInterests(): Observable<string[]> {
    return this.http
      .get<ITopic[]>(`${API_END_POINTS.userTopics}?ts=${new Date().getTime()}`)
      .pipe(map(u => u.map((topic: ITopic) => topic.name)))
  }

  fetchSuggestedInterests(): Observable<string[]> {
    return this.http
      .get<ITopicRecommended[]>(`${API_END_POINTS.recommendedTopics}`)
      .pipe(map(u => u.map((topic: ITopicRecommended) => topic.name)))
  }

  modifyUserInterests(topics: string[]) {
    return this.http.post(API_END_POINTS.userTopics, { topics })
  }

  fetchAutocompleteInterests(): Observable<string[]> {
    const leafNodes: NSSearch.IFilterUnitContent[] = []
    return this.treeCatalogSvc.getFullCatalog().pipe(
      map(catalogDataTree => {
        catalogDataTree.forEach(catalogUnit => {
          leafNodes.push(...this.getLeafNodes(catalogUnit, []))
        })
        return leafNodes.map(unitItem => unitItem.displayName)
      }),
    )
  }
  getLeafNodes<T extends NSSearch.IFilterUnitContent>(node: T, nodes: T[]): T[] {
    if ((node.children || []).length === 0) {
      nodes.push(node)
    } else {
      if (node.children) {
        node.children.forEach(child => {
          this.getLeafNodes(child, nodes)
        })
      }
    }
    return nodes
  }

  // V2 Interest code
  fetchUserInterestsV2(): Observable<string[]> {
    return this.http.get<string[]>(`${API_END_POINTS.userTopics}/v2?ts=${new Date().getTime()}`)
  }

  addUserInterest(interest: string) {
    return this.http.patch(`${API_END_POINTS.userTopics}?interest=${interest}`, {})
  }

  removeUserInterest(interest: string) {
    return this.http.delete(`${API_END_POINTS.userTopics}?interest=${interest}`)
  }

  fetchSuggestedInterestV2(): Observable<string[]> {
    return this.http.get<string[]>(`${API_END_POINTS.recommendedTopics}`)
  }

  fetchAutocompleteInterestsV2(query: string): Observable<string[]> {
    return this.http.get<string[]>(`${API_END_POINTS.autocompleteTopics}?query=${query}`)
  }
}
